From e2b58cb2eab53c07f921a9b8d1006f2dfb3ebcdc Mon Sep 17 00:00:00 2001
From: Amith A <amitajit@qti.qualcomm.com>
Date: Wed, 25 Jun 2025 20:09:35 +0530
Subject: [PATCH] QSDK:PROTOCOL: wifi: ath12k: Add an ath12k_ops to get the
 6 GHz device deployment type

Add an ath12k_ops, 'get_6ghz_dev_deployment_type' to get the 6 GHz device
deployment type.

Fixes: None
Patch-work: None
Tag: 6G_REGULATORY
Feature: AFC
Path-dependency:
subsys/0332-04-UPSTREAM-PROTOCOL-wifi-mac80211-Add-a-mac80211_config_op.patch 
0375-UPSTREAM-PROTOCOL-wifi-ath12k-Add-module-params-for-.patch

Signed-off-by: Amith A <amitajit@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 32 +++++++++++++++++++++++++++
 drivers/net/wireless/ath/ath12k/mac.h |  3 +++
 2 files changed, 35 insertions(+)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -22053,3 +22053,39 @@ int ath12k_mac_op_get_afc_eirp_pwr(struc
 	return 0;
 }
 EXPORT_SYMBOL(ath12k_mac_op_get_afc_eirp_pwr);
+
+int ath12k_mac_op_get_6ghz_dev_deployment_type(struct ieee80211_hw *hw,
+					       enum nl80211_6ghz_dev_deployment_type *dep_type)
+{
+	struct ath12k_hw *ah = hw->priv;
+	struct ath12k_base *ab;
+	struct ath12k *ar = NULL;
+	int i;
+
+	for_each_ar(ah, ar, i) {
+		if (ar->supports_6ghz)
+			break;
+	}
+
+	if (!ar) {
+		ath12k_err(NULL, "ar is NULL\n");
+		return -ENODATA;
+	}
+
+	ab = ar->ab;
+	switch (ab->afc_dev_deployment) {
+	case ATH12K_AFC_DEPLOYMENT_INDOOR:
+		*dep_type = NL80211_6GHZ_DEV_DEPLOYMENT_TYPE_INDOOR;
+		break;
+	case ATH12K_AFC_DEPLOYMENT_OUTDOOR:
+		*dep_type = NL80211_6GHZ_DEV_DEPLOYMENT_TYPE_OUTDOOR;
+		break;
+	case ATH12K_AFC_DEPLOYMENT_UNKNOWN:
+	default:
+		*dep_type = NL80211_6GHZ_DEV_DEPLOYMENT_TYPE_UNKNOWN;
+		break;
+	}
+
+	return 0;
+}
+EXPORT_SYMBOL(ath12k_mac_op_get_6ghz_dev_deployment_type);
--- a/drivers/net/wireless/ath/ath12k/mac.h
+++ b/drivers/net/wireless/ath/ath12k/mac.h
@@ -574,4 +574,12 @@ int ath12k_mac_op_qos_mgmt_cfg(struct ie
  */
 int ath12k_mac_op_get_afc_eirp_pwr(struct ieee80211_hw *hw, u32 freq,
 				   u32 *eirp_pwr);
+/**
+ * ath12k_mac_op_get_6ghz_dev_deployment_type - Get the 6 GHz device deployment
+ * type.
+ * @hw: Input hardware
+ * @dep_type: Output deployment type
+ */
+int ath12k_mac_op_get_6ghz_dev_deployment_type(struct ieee80211_hw *hw,
+					       enum nl80211_6ghz_dev_deployment_type *dep_type);
 #endif
--- a/drivers/net/wireless/ath/ath12k/wifi7/hw.c
+++ b/drivers/net/wireless/ath/ath12k/wifi7/hw.c
@@ -1697,6 +1697,7 @@ static const struct ieee80211_ops ath12k
 	.set_radar_background           = ath12k_mac_op_set_radar_background,
 	.qos_mgmt_cfg			= ath12k_mac_op_qos_mgmt_cfg,
 	.get_afc_eirp_pwr               = ath12k_mac_op_get_afc_eirp_pwr,
+	.get_6ghz_dev_deployment_type	= ath12k_mac_op_get_6ghz_dev_deployment_type,
 };
 
 int ath12k_wifi7_hw_init(struct ath12k_base *ab)
